## Programa Ingenias+ Data Science

## Trabajando con Pandas

1) Importa las librerias requeridas: `Pandas` y `NumPy`.  

Una buena practica a la hora de escribir codigo es de comentar siempre que estoy haciendo (usando `#`) en la linea anterior. Por mas obvio que parezca lo que voy a hacer, puede ser que tenga que compartir el documento con otras personas y no sepan que quise hacer en cada linea. O que yo mismo lea el codigo tiempo despues y no recuerde porque lo hice.

In [1]:
# Importo librerias y renombramos para usarlas
import numpy as np
import pandas as pd

2) Lee el archivo csv `life_expectancy_data` usando `pandas`. Guardalo en una variable llamada `expvida`.

In [2]:
# Leo los datos en un DataFrame
expvida= pd.read_csv('LifeExpectancyData.csv')

3) ¿Que tipo de estructura de datos contiene la variable `expvida`?

In [5]:
# Obtengo el tipo de estructura de datos usando función type
print(f'El tipo de estructura de los datos en expvida', type(expvida))

El tipo de estructura de los datos en expvida <class 'pandas.core.frame.DataFrame'>


4) ¿Cuantas filas y columnas tiene `expvida`?

In [6]:
# Uso .shape para obtener la dimensión del DataFrame (cantidad de filas y columnas)
filas_columnas = expvida.shape
print(f'El DataFrame tiene {filas_columnas[0]} filas y {filas_columnas[1]} columnas' )

El DataFrame tiene 2938 filas y 22 columnas


**Numero de filas**: 2938

**Numero de columnas**: 22

5) ¿Cual es el nombre de las columnas contenidas en `expvida`?

In [7]:
# Obtengo el nombre de las columnas con .columns
print(f'El nombre de las columnas es', expvida.columns)

El nombre de las columnas es Index(['Country', 'Year', 'Status', 'Life expectancy ', 'Adult Mortality',
       'infant deaths', 'Alcohol', 'percentage expenditure', 'Hepatitis B',
       'Measles ', ' BMI ', 'under-five deaths ', 'Polio', 'Total expenditure',
       'Diphtheria ', ' HIV/AIDS', 'GDP', 'Population',
       ' thinness  1-19 years', ' thinness 5-9 years',
       'Income composition of resources', 'Schooling'],
      dtype='object')


6) Inspecciona las primeras 10 filas de `expvida`

In [8]:
# Inspecciono las primeras filas con .head
print(expvida.head(10))

       Country  Year      Status  Life expectancy   Adult Mortality  \
0  Afghanistan  2015  Developing              65.0            263.0   
1  Afghanistan  2014  Developing              59.9            271.0   
2  Afghanistan  2013  Developing              59.9            268.0   
3  Afghanistan  2012  Developing              59.5            272.0   
4  Afghanistan  2011  Developing              59.2            275.0   
5  Afghanistan  2010  Developing              58.8            279.0   
6  Afghanistan  2009  Developing              58.6            281.0   
7  Afghanistan  2008  Developing              58.1            287.0   
8  Afghanistan  2007  Developing              57.5            295.0   
9  Afghanistan  2006  Developing              57.3            295.0   

   infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles   ...  \
0             62     0.01               71.279624         65.0      1154  ...   
1             64     0.01               73.523582       

7) Y ahora las 10 ultimas.

In [9]:
# Inspecciono las ultimas filas con .tail
print(expvida.tail(10))

       Country  Year      Status  Life expectancy   Adult Mortality  \
2928  Zimbabwe  2009  Developing              50.0            587.0   
2929  Zimbabwe  2008  Developing              48.2            632.0   
2930  Zimbabwe  2007  Developing              46.6             67.0   
2931  Zimbabwe  2006  Developing              45.4              7.0   
2932  Zimbabwe  2005  Developing              44.6            717.0   
2933  Zimbabwe  2004  Developing              44.3            723.0   
2934  Zimbabwe  2003  Developing              44.5            715.0   
2935  Zimbabwe  2002  Developing              44.8             73.0   
2936  Zimbabwe  2001  Developing              45.3            686.0   
2937  Zimbabwe  2000  Developing              46.0            665.0   

      infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles   \
2928             30     4.64                1.040021         73.0       853   
2929             30     3.56               20.843429        

8) ¿Que tipos de datos contiene cada una de las columnas de `expvida`?

In [12]:
# Tipo de datos en cada columna usando .dtypes
print('Los tipos de datos para cada columna son:')
print(expvida.dtypes)

Los tipos de datos para cada columna son:
Country                             object
Year                                 int64
Status                              object
Life expectancy                    float64
Adult Mortality                    float64
infant deaths                        int64
Alcohol                            float64
percentage expenditure             float64
Hepatitis B                        float64
Measles                              int64
 BMI                               float64
under-five deaths                    int64
Polio                              float64
Total expenditure                  float64
Diphtheria                         float64
 HIV/AIDS                          float64
GDP                                float64
Population                         float64
 thinness  1-19 years              float64
 thinness 5-9 years                float64
Income composition of resources    float64
Schooling                          float64
dtype: objec

9) ¿Hay valores faltantes en alguna de las columnas de `expvida`?

In [13]:
# Obtengo la cantidad de datos faltantes en cada columna usando .isnull para encontrar los nulos
# y .sum para cuantificarlos en cada columna
print('La cantidad de datos faltantes para cada columna es:')
expvida.isnull().sum() 

La cantidad de datos faltantes para cada columna es:


Country                              0
Year                                 0
Status                               0
Life expectancy                     10
Adult Mortality                     10
infant deaths                        0
Alcohol                            194
percentage expenditure               0
Hepatitis B                        553
Measles                              0
 BMI                                34
under-five deaths                    0
Polio                               19
Total expenditure                  226
Diphtheria                          19
 HIV/AIDS                            0
GDP                                448
Population                         652
 thinness  1-19 years               34
 thinness 5-9 years                 34
Income composition of resources    167
Schooling                          163
dtype: int64

10) Selecciona el maximo de datos faltantes. Calcula que porcentaje de los datos representan los datos faltantes.

In [15]:
# Obtengo el mayor número de valores nulos y lo guardo en la variable maxnan
maxnan=expvida.isnull().sum().max()
print(f'El máximo valor de datos faltantes en una columna es', maxnan)

El máximo valor de datos faltantes en una columna es 652


In [16]:
# Obtengo la cantidad de filas total y divido el maximo valor de datos faltantes por este numero
# y lo multiplico por 100

#la cantidad de filas total es
f = len(expvida)
print(f'La cantidad de filas del DataFrame es',f)

#obtengo porcentaje que ocupan los datos faltantes en relación al total de filas
maxp = (maxnan*100)/f

La cantidad de filas del DataFrame es 2938


In [17]:
# Imprimo el maximo porcentaje de datos faltantes. Uso f-strings.
print(f'El porcentaje que ocupan los datos faltantes en el DataFrame es {maxp:.1f}%')

El porcentaje que ocupan los datos faltantes en el DataFrame es 22.2%


- str.format()
- Template Class
- f-strings

#### Ahora vamos a realizar algunos calculos y manipulaciones sobre las columnas.

Una función útil para este fin es: `.rename()`. Y podemos utilizarla de la siguiente manera: Si `df` es mi DataFrame y quiero renombrar la columna `columna` como `columna1`, y la columna `columna2` como `columna3`, utilizo la siguiente sintaxis:
  
`df.rename(columns={"columna": "columna1", "columna2": "columna3"}, inplace=True)`

11) Renombra la columna 'Life expectancy ' como 'life_expectancy', ' BMI ' como 'bmi' y ''Measles ' como 'measles'. No olvides de usar `inplace` para modificar el DataFrame original. Luego chequea que el cambio se haya hecho correctamente.

In [18]:
# Chequeo cómo aparecen escritos los nombres en cada columna
print('Los nombres actuales de cada columna son:')
print(expvida.columns.tolist())

Los nombres actuales de cada columna son:
['Country', 'Year', 'Status', 'Life expectancy ', 'Adult Mortality', 'infant deaths', 'Alcohol', 'percentage expenditure', 'Hepatitis B', 'Measles ', ' BMI ', 'under-five deaths ', 'Polio', 'Total expenditure', 'Diphtheria ', ' HIV/AIDS', 'GDP', 'Population', ' thinness  1-19 years', ' thinness 5-9 years', 'Income composition of resources', 'Schooling']


In [19]:
# Renombro columnas del DataFrame según consigna usando función rename
expvida.rename(columns={"Life expectancy ": "life_expectancy", " BMI ": "bmi", "Measles ":"measles"}, inplace=True)

In [20]:
# Chequeo que se hayan reemplazado correctamente
print('Cambiando los nombres que se requieren según la consigna queda:')
print(expvida.columns.tolist())

Cambiando los nombres que se requieren según la consigna queda:
['Country', 'Year', 'Status', 'life_expectancy', 'Adult Mortality', 'infant deaths', 'Alcohol', 'percentage expenditure', 'Hepatitis B', 'measles', 'bmi', 'under-five deaths ', 'Polio', 'Total expenditure', 'Diphtheria ', ' HIV/AIDS', 'GDP', 'Population', ' thinness  1-19 years', ' thinness 5-9 years', 'Income composition of resources', 'Schooling']


12) ¿Cual es el promedio de expectativa de vida (Life expectancy) en general?

In [21]:
# Obtengo el promedio de la columna Life_expectancy usando función mean y renombrando el valor
mexp = expvida['life_expectancy'].mean()
# Muestro en pantalla el resultado redondeado a 1 decimal
print(f'El valor medio de la expectativa de vida general es de {mexp:.1f} años')

El valor medio de la expectativa de vida general es de 69.2 años


13) ¿Cual es el promedio de expectativa de vida (Life expectancy) por país?

In [22]:
# Obtengo el promedio por pais y redondeo a 1 decimal
prom = expvida.groupby('Country', as_index=False)['life_expectancy'].mean().round(1)
print("El promedio de la expectativa de vida por país es")
print(prom)

El promedio de la expectativa de vida por país es
                                Country  life_expectancy
0                           Afghanistan             58.2
1                               Albania             75.2
2                               Algeria             73.6
3                                Angola             49.0
4                   Antigua and Barbuda             75.1
..                                  ...              ...
188  Venezuela (Bolivarian Republic of)             73.4
189                            Viet Nam             74.8
190                               Yemen             63.9
191                              Zambia             53.9
192                            Zimbabwe             50.5

[193 rows x 2 columns]


14) ¿Cual es la expectativa de vida mas alta y mas baja? Usa la funcion de pandas y la funcion de NumPy y compara el resultado. ¿Que paises tienen esas expectativas de vida?

In [23]:
# Maxima expectativa de vida usando numpy
max_numpy = np.max(expvida["life_expectancy"])

In [24]:
# Maxima expectativa de vida usando pandas
max_pandas = expvida["life_expectancy"].max()

In [25]:
# Minima expectativa de vida usando numpy
min_numpy = np.min(expvida["life_expectancy"])

In [26]:
# Minima expectativa de vida usando pandas
min_pandas = expvida["life_expectancy"].min()

In [27]:
# Muestro los resultados maximos
print(f'La máxima expectativa de vida es de {max_numpy} años')
print(f'La máxima expectativa de vida es de {max_pandas} años')

##los valores son iguales
## la clave está en saber que existen estas dos maneras de calcular


La máxima expectativa de vida es de 89.0 años
La máxima expectativa de vida es de 89.0 años


In [28]:
# Muestro los resultados minimos
print(f'La minima expectativa de vida es de {min_numpy} años')
print(f'La minima expectativa de vida es de {min_pandas} años')

La minima expectativa de vida es de 36.3 años
La minima expectativa de vida es de 36.3 años


In [44]:
# Filtro las columnas para mostrar solo las filas que tengan valor
# igual al maximo
filas_max = expvida[expvida["life_expectancy"] == max_pandas] 
# filtro para que aparezcan una sola vez los nombre de esos paises
maxun = filas_max['Country'].unique()
# toma los elementos del array maxun y los une en un solo string, separados por coma
print(f'El/Los paises que tienen la mayor expectativa de vida son: {", ".join(maxun)}')

El/Los paises que tienen la mayor expectativa de vida son: Belgium, Finland, France, Germany, Italy, New Zealand, Norway, Portugal, Spain, Sweden


In [45]:
# Filtro las columnas para mostrar solo las filas que tengan valor
# igual al minimo
filas_min = expvida[expvida["life_expectancy"] == min_pandas]
# filtro para que aparezcan una sola vez los nombre de esos paises
minun = filas_min['Country'].unique()
print(f'El/Los paises que tienen la menor expectativa de vida son: {", ".join(minun)}')


El/Los paises que tienen la menor expectativa de vida son: Haiti


15) ¿Cuantas categorias encontramos en la columna "Status"? _**Pista**_: Lee acerca de la función [unique()](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.unique.html?highlight=unique#pandas.Series.unique)

In [49]:
# Obtengo los valores unicos de la columna status
stun = expvida['Status'].unique()
# mostramos cuántas categorias encontramos en la columna Status
print(f"En la columna Status hay {len(stun)} categorías: {', '.join(stun)}")

En la columna Status hay 2 categorías: Developing, Developed


16) ¿Cuantos paises se ubican en cada una de esas categorias de "Status"? _**Pista**_: Lee acerca de la función [value_counts()](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.value_counts.html)

In [53]:
# Cuento cuantos valores hay de cada categoria
st = expvida['Status'].value_counts()
for categoria, cantidad in st.items():
    print(f"En la categoría '{categoria}' hay {cantidad} valores")

En la categoría 'Developing' hay 2426 valores
En la categoría 'Developed' hay 512 valores


18) ¿Cuantos paises presentaron algun caso de Sarampion (**Measles**) en el 2015? **_Pista_**: Lee acerca de la función [.count()](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.count.html?highlight=count#pandas.DataFrame.count)

In [56]:
paises_con_sarampion_2015 = expvida[(expvida['Year'] == 2015) & (expvida['measles'] > 0)]
cantidad_paises = paises_con_sarampion_2015['Country'].count()
#otra opción puede ser usar .unique, para que cuente solo 1 vez cada país
#en este ejemplo, da lo mismo
#cantidad_paises = paises_con_sarampion_2015['Country'].nunique()
print(f"Cantidad de países con casos de sarampión en 2015: {cantidad_paises}")


Cantidad de países con casos de sarampión en 2015: 121
